new_ux.py

```diff
@@ -0,0 +1,186 @@
+# function parameter type annotation if it has a constructor from dict() that we deserialized from json.
+#
+# API function must have a single user supplied argument caller `request`, this is HTTP calling convention. If the first argument
+# is called `ctx` then RequestContext is injected. User has to pass it to all functions that it calls if they have ctx parameter too,
+# This is mainly to make this code look like Python and thus get all the IDE and LLM support smoothly.
+#
+# The decorator attributes are the same as @tensorlake.function + attributes for defining the graph.
+@tensorlake.graph_api(graph_name="test_graph", graph_description="test", version="2.0")
+def test_graph_function_1_api(
+    ctx: tensorlake.RequestContext, request: dict
+) -> tensorlake.FunctionCall:  # Actual return type, all return types are ignored by SDK
+    print(
+        f"Received request with key1: {request.get('key1')}, key2: {request.get('key2')}"
+    )
+    ctx.state.set("key1", request.get("key1"))
+    # Require explicit passing of the context, this aligns with its semantic that it's the same context per request.
+    return test_graph_function_2_map(ctx, request, times=10)
+
+
+# Non API functions can have multiple arguments
+@tensorlake.function()
+def test_graph_function_2_map(
+    ctx: tensorlake.RequestContext, d: dict, times: int
+) -> Iterable[
+    tensorlake.FunctionCall
+]:  # The returned list can have both function calls and outputs for this function, no problem with that.
+    print(
+        f"Processing data in test_graph_function_2 with key1: {d.get('key1')}, key2: {d.get('key2')}, times: {times}"
+    )
+    return [TestGraphFunction3().run(ctx, d, ix) for ix in range(times)]
+    # Or tuple:
+    # return (TestGraphFunction3().run(ctx, d, 0), TestGraphFunction3().run(ctx, d, 1), TestGraphFunction3().run(ctx, d, 2))
+    # Or map:
+    # return map(lambda ix: TestGraphFunction3().run(ctx, d, ix), range(times))
+
+
+# When @tensorlake.function() is applied to a class it uses its __init__(self) as one time container initialization hook.
+# The class must have def run(self, ...) method. This is the actual function body.
+@tensorlake.function()
+# Also available:
+# @tensorlake.reducer(...)
+# @tensorlake.graph_api(...)
+class TestGraphFunction3:
+    # The class must have an empty constructor.
+    # This is because when this class is called by other functions using `TestGraphFunction3().run(...)` we don't support
+    # different constructor arguments for different requests. In the future we can support it by e.g. hashing constructor
+    # args and kwargs and adding this to the function name but this is too complex to support right now.
+
+    # The decorator makes actual class constructor body empty when ppl call TestGraphFunction3(). When however we're running
+    # a graph the original __init__ is actually called.
+    def __init__(self):
```

Comment:
During implementation we should validate if these constructors take more than one argument.

Reply:
Yes, because `class` is a keyword. I don't have any other ideas for the name, if you have some the please share. The only other alternative I have is `tensorlake.function_class` but this looks like mixing terms to me which causes more harm than improvement here imo.